### PR TITLE
implements have name start with and ends with class expressions

### PR DESCRIPTION
**Pull request by fokosun**
### What does this PR do?

Implements two new class expressions namely:  HaveNameEndingWith and  HaveNameStartingWith

### HaveNameEndingWith
A class expression that evaluates that the classname ends with a given string

### HaveNameStartingWith
A class expression that evaluates that the classname starts with a given string

### CODE REVIEW

**Comment by pfazzi** [comment]
Hello @fokosun, thanks for contributing :)

At the beginning of this project we have been debating about the minimum PHP version to support and we finally agreed to maintain backward compatibility with older PHP versions, since PHP 7.1. The reason is that we want to allow people anchored in ancient PHP versions to install and run the tool easily. 

Please, could you use `strpos`  instead of new functions in order to guarantee compatibility with older PHP versions? Thansk 🙂

**Comment by fokosun** [comment]
> Hello @fokosun, thanks for contributing :)
> 
> At the beginning of this project we have been debating about the minimum PHP version to support and we finally agreed to maintain backward compatibility with older PHP versions, since PHP 7.1. The reason is that we want to allow people anchored in ancient PHP versions to install and run the tool easily.
> 
> Please, could you use `strpos` instead of new functions in order to guarantee compatibility with older PHP versions? Thansk 🙂

got it! working on that. Thanks for the feedback

**Comment by pfazzi** [comment]
@fokosun you should comply to coding styles. Please run `make build` before to push ;)

File: src/Expression/ForClasses/HaveNameEndsWith.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+use Arkitect\Analyzer\ClassDescription;
+use Arkitect\Analyzer\FullyQualifiedClassName;
+use Arkitect\Expression\Description;
+use Arkitect\Expression\Expression;
+use Arkitect\Expression\PositiveDescription;
+use Arkitect\Rules\Violation;
+use Arkitect\Rules\Violations;
+
+class HaveNameEndsWith implements Expression
```
Comment (pfazzi) [inline src/Expression/ForClasses/HaveNameEndsWith.php:14]: English is not my mother language,  but let me ask the same.  Isn't `HaveNameEndingWith` more correct than `HaveNameEndsWith`?

File: src/Expression/ForClasses/HaveNameStartsWith.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    public function describe(ClassDescription $theClass): Description
+    {
+        return new PositiveDescription("should have a name that starts with {$this->name}");
+    }
+
+    public function evaluate(ClassDescription $theClass, Violations $violations): void
+    {
+        $fqcn = FullyQualifiedClassName::fromString($theClass->getFQCN());
+
+        if (!(0 === strpos($fqcn->className(), $this->name))) {
```
Comment (pfazzi) [inline src/Expression/ForClasses/HaveNameStartsWith.php:33]: Can you simplify this condition? `0 !== strpos($fqcn->className(), $this->name)` seems more readable :)

File: tests/Unit/Expressions/ForClasses/HaveNameEndsWithTest.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public function test_show_violation_when_class_name_does_not_end_with(): void
+    {
+        $expression = new HaveNameEndsWith('Suffix');
+
+        $badClass = ClassDescription::build('\App\BadNameClass')->get();
+
+        $violations = new Violations();
+        $expression->evaluate($badClass, $violations);
+        self::assertNotEquals(0, $violations->count());
+        $this->assertEquals('should have a name that ends with Suffix', $expression->describe($badClass)->toString());
```
Comment (pfazzi) [inline tests/Unit/Expressions/ForClasses/HaveNameEndsWithTest.php:33]: It's better to be consistent with PHPUnit method calls. Use the static syntax in this case ;)

**Review by pfazzi — CHANGES_REQUESTED** [review]

**Comment by AlessandroMinoccheri** [comment]
@fokosun thanks for this PR!
I agree with @pfazzi comments and @fokosun, if you need some helps let me know.

File: tests/Unit/Expressions/ForClasses/HaveNameEndsWithTest.php (hunk 3 above)
Comment (fokosun) [reply tests/Unit/Expressions/ForClasses/HaveNameEndsWithTest.php]: Done

File: src/Expression/ForClasses/HaveNameStartsWith.php (hunk 2 above)
Comment (fokosun) [reply src/Expression/ForClasses/HaveNameStartsWith.php]: Done

File: src/Expression/ForClasses/HaveNameEndingWith.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    public function describe(ClassDescription $theClass): Description
+    {
+        return new PositiveDescription("should have a name that ends with {$this->name}");
+    }
+
+    public function evaluate(ClassDescription $theClass, Violations $violations): void
+    {
+        $fqcn = FullyQualifiedClassName::fromString($theClass->getFQCN());
+
+        $sub_str = substr($fqcn->className(), (-1 * strlen($this->name)));
```
Comment (pfazzi) [inline src/Expression/ForClasses/HaveNameEndingWith.php:33]: Variable naming should be in camelCase :)

**Review by pfazzi — CHANGES_REQUESTED** [review]

File: src/Expression/ForClasses/HaveNameEndingWith.php (hunk 4 above)
Comment (fokosun) [reply src/Expression/ForClasses/HaveNameEndingWith.php:33]: Apologies for the delay in submitting the requested change. I'll be pushing up the change soon. Thanks

**Comment by diimpp** [comment]
Is there a reason to introduce new expressions, when `HaveNameMatching('*Controller')` already exists?

**Comment by fokosun** [comment]
> Is there a reason to introduce new expressions, when `HaveNameMatching('*Controller')` already exists? If wildcard is not good enough, then perhaps generic regexp expression can be added

You have a valid point there! Apologies for the long delay in responding and working to address the issues on this PR. It looks like this PR is not really necessary then based on this comment. Is it okay to close the PR?

**Comment by AlessandroMinoccheri** [comment]
Ok @fokosun I close this PR but feel free to contribute again :)

**Pull request closed without merging**